helix-term/src/commands.rs

```diff
@@ -1076,16 +1077,22 @@ mod cmd {
             .language_config()
             .map(|config| config.auto_format)
             .unwrap_or_default();
-        if autofmt {
-            doc.format(view.id); // TODO: merge into save
-        }
-        helix_lsp::block_on(tokio::spawn(doc.save()));
+        let fmt = if autofmt {
+            doc.format().map(|fmt| {
+                let shared = fmt.shared();
+                let callback = make_format_callback(doc.id(), doc.version(), true, shared.clone());
+                callbacks.push(callback);
+                shared
+            })
+        } else {
+            None
+        };
+        tokio::spawn(doc.format_and_save(fmt));
```

Comment:
The `block_on` was necessary here. It's a bit of a hack but without it the editor could quit before a file save completed.

Reply:
Ah, that's a problem. Because if we're blocking until the save is done, then we're also blocking until the formatting is done and so this PR won't achieve anything...

What if we were to store the `JoinHandle`s from spawning the save job, and make sure to block on them before exit?

Reply:
I was thinking that a more general version of the `LspCallbacks` would be useful (not limited to LSP). We could spawn the futures there, then receive the values. If we get a `Error` return we can also log it to the `Editor` via `set_error`. This would solve the problem of not being able to report errors from these utility tasks.

The shutdown sequence would then fully drain this queue (or time out) before dropping the runtime.

Do you want to make the change in this PR? I can probably look into it otherwise.

(By the way, feel free to join us on Matrix! we were just discussing your teddy library earlier)

Reply:
Sure, I can take a stab at this. I'm not 100% sure what you mean by "receive the values", because I would expect that different jobs would want to return different things, and then how would the editor know how to handle them? I guess what I'm trying to say is that I don't see a way to completely avoid the callback...

I was imagining a `AsyncJob` struct (or maybe trait?) that contains
- a `Future` returning an `anyhow::Result<Callback>`
- a `JoinHandle` for cancelling that future
- some additional metadata, like timeouts, and whether we need to wait on the future before exiting

Does that sound reasonable?

Reply:
I was imagining something similar, but simpler:

`lsp_callbacks` is currently defined as `FuturesUnordered<LspCallback>`. I'd turn that into a `callbacks: FuturesUnordered<Result<(), anyhow::Error>>`. The lsp futures will then need to be wrapped in another future to work as before.

`cx.spawn(doc.save)` would push the future onto the callbacks. Since the callback results are received on in the main event loop here: https://github.com/helix-editor/helix/blob/82fc28a0ce2cb2b86968335f32e95aaff3e2cba4/helix-term/src/application.rs#L175-L177

We could catch any `Error`s here and use `editor.set_error` to show them ("save failed" etc).

When quitting we just need to drain the `callbacks` stream until empty or a timeout is reached.

Reply:
> (By the way, feel free to join us on Matrix! we were just discussing your teddy library earlier)

Sure, I'll join! It's really BurntSushi's library, though. I just extracted it into a crate so I could play with it...

Reply:
>  I'd turn that into a callbacks: `FuturesUnordered<Result<(), anyhow::Error>>`. The lsp futures will then need to be wrapped in another future to work as before.

How would you get access to the `&mut Editor` in the wrapper callback, though? I thought that borrowing issue was the reason for the whole callback pattern.